src/python/pants/engine/internals/rule_visitor.py

```diff
@@ -21,11 +21,12 @@
     GetParseError,
     MultiGet,
 )
-from pants.util.backport import get_annotations
 from pants.util.memo import memoized
 from pants.util.strutil import softwrap
+from pants.util.typing import patch_forward_ref
 
 logger = logging.getLogger(__name__)
+patch_forward_ref()
```

Comment:
We never call restore_forward_ref() so we never undo this monkeypatch.  Probably not an issue? I assume we have no control over when the typing is evaluated, so we can't constrain the monkeypatch to just that context?

Reply:
hmm.. actually I think we do. This needs to be in place only for the duration of loading all rule code. Once the graph has been created we're done and can undo this monkey business. I think. But I'm not sure if it's a big issue at leave it any way.. Maybe wait and see?